Description: amd64_fixes.patch
Author: Nick Bane <nick@enomem.co.uk>

--- a/src/numutils.c
+++ b/src/numutils.c
@@ -12,18 +12,18 @@
  * way
  *********************************************************************/ 
 uint16_t getshort(unsigned char *buffer,int offset) {
-	return (unsigned short int)buffer[offset]|((unsigned short int)buffer[offset+1]<<8);
+	return (uint16_t)buffer[offset]|((uint16_t)buffer[offset+1]<<8);
 }  
 /********************************************************************/
 /* Reads 4-byte LSB  int from buffer at given offset almost platfom-indepent
  * way
  *********************************************************************/ 
 int32_t getlong(unsigned char *buffer,int offset) {
-	return (long)buffer[offset]|((long)buffer[offset+1]<<8L)
-		|((long)buffer[offset+2]<<16L)|((long)buffer[offset+3]<<24L);
+	return (int32_t)buffer[offset]|((int32_t)buffer[offset+1]<<8L)
+		|((int32_t)buffer[offset+2]<<16L)|((int32_t)buffer[offset+3]<<24L);
 }  
 
 uint32_t getulong(unsigned char *buffer,int offset) {
-	return (unsigned long)buffer[offset]|((unsigned long)buffer[offset+1]<<8L)
-		|((unsigned long)buffer[offset+2]<<16L)|((unsigned long)buffer[offset+3]<<24L);
+	return (uint32_t)buffer[offset]|((uint32_t)buffer[offset+1]<<8L)
+		|((uint32_t)buffer[offset+2]<<16L)|((uint32_t)buffer[offset+3]<<24L);
 }  
